cleverhans/attacks_eager.py

```diff
@@ -0,0 +1,260 @@
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model, dtypestr='float32'):
+        """
+        :param model: An instance of the cleverhans.model.Model_Eager class.
+        :param back: The backend to use. Inherited from AttackBase class.
+        :param dtypestr: datatype of the input data samples and crafted
+                        adversarial attacks.
+        """
+        # Validate the input arguments.
+        if dtypestr != 'float32' or dtypestr != 'float64':
+            raise ValueError("Unexpected input for argument dtypestr.")
+        import tensorflow as tf
+        tfe = tf.contrib.eager
+        self.tf_dtype = tf.as_dtype(dtypestr)
+        self.np_dtype = np.dtype(dtypestr)
+
+        if not isinstance(model, Model_Eager):
+            raise ValueError("The model argument should be an instance of"
+                             " the cleverhans.model.Model_Eager class.")
+        # Prepare attributes
+        self.layer_names = []
+
+    def get_or_guess_labels(self, x, kwargs):
+        """
+        Get the label to use in generating an adversarial example for x.
+        The kwargs are fed directly from the kwargs of the attack.
+        If 'y' is in kwargs, then assume it's an untargeted attack and
+        use that as the label.
+        If 'y_target' is in kwargs and is not none, then assume
+        it's a targeted attack and use that as the label.
+        Otherwise, use the model's prediction as the label and perform an
+        untargeted attack.
+        :param x: batch of input images, as tensorflow variable
+        :return: labels for crafting attack based on y and y_target.
+        """
+        if 'y' in kwargs and 'y_target' in kwargs:
+            raise ValueError("Can not set both 'y' and 'y_target'.")
+        elif 'y' in kwargs:
+            labels = kwargs['y']
+        elif 'y_target' in kwargs and kwargs['y_target'] is not None:
+            labels = kwargs['y_target']
+        else:
+            preds = self.model(x)
+            preds_max = tf.reduce_max(preds, 1, keep_dims=True)
+            original_predictions = tf.to_float(tf.equal(preds,
+                                                        preds_max))
+            labels = tf.stop_gradient(original_predictions)
+        if isinstance(labels, np.ndarray):
+            nb_classes = labels.shape[1]
+        else:
+            nb_classes = labels.get_shape().as_list()[1]
+        return labels, nb_classes
+
+    def generate_np(self, x_val, **kwargs):
+        """
+        Generate adversarial examples and return them as a NumPy array.
+
+        :param x_val: A NumPy array with the original inputs.
+        :param **kwargs: optional parameters used by child classes.
+        :return: A NumPy array holding the adversarial examples.
+        """
+        tfe = tf.contrib.eager
+        x = tfe.Variable(x_val)
+        adv_x = self.generate(x, **kwargs)
+        return adv_x.numpy()
+
+    def clip_eta(self, eta, ord, eps):
```

Comment:
What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

Reply:
Same as before, lets discuss in the previous comment.